README.rst
============================
``gs.group.member.add.base``
============================
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Add a member to a GroupServer Group
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

:Author: `Michael JasonSmith`_
:Contact: Michael JasonSmith <mpj17@onlinegroups.net>
:Date: 2015-03-03
:Organization: `GroupServer.org`_
:Copyright: This document is licensed under a
  `Creative Commons Attribution-Share Alike 4.0 International License`_
  by `OnlineGroups.net`_.

Introduction
============

This product provides the "just add" system for
GroupServer_. Normally people become participants by actively
accepting an invitation [#invite]_, signing up [#register]_, or
joining a group [#join]_. Generally these methods are better at
getting people into a group than the `Add Form`_. People who are
added have no password, so they cannot use any of the features on
the Web. They are added are dependent on the administrator for
all help. In addition all the help pages, all the email footers,
and most of the notifications are written assuming that the
member can log in.

Add Form
========

The *Add* form is very similar to the *Invite a New Member*
form. It inherits much of the functionality from the Invite
system. Like the Invite system, it requires two email addresses:
a ``From`` address and a ``To`` address. The latter is used as
the email address of the new member. All the required profile
fields have to be filled out, also. There are three possible
outcomes from submitting the Add form.

New Profile:
  If the email address is unrecognised then a new profile is
  created, the member is added to the group, and he or she
  receives the `Welcome notification`_.

Existing Profile:
  If the email address already belongs to an existing profile,
  and that participant is not a member of the group, then he or
  she is joined to the group and sent the *standard* Welcome
  notification from ``gs.group.member.join`` [#join]_. The
  (somewhat vein) hope is that the participant would have set a
  password and is able to log in by the time the new member is
  added to a second group.
  
Existing Member:
  If the email address already belongs to an existing group
  member then no action is taken.

The *Email delivery settings* selector is not present on the Add
form. This is because the new member cannot login [#login]_, so
he or she generally cannot view the posts on the web, which is
necessary for *Web only* and the *Topic digest* options to work
[#digest]_.

Welcome Notification
--------------------

The *Welcome* notification sent from this product contains no
links to the group or the profile of the member. This is because
*generally* the member cannot view anything without logging in
[#login]_, and the member is unable to log in without a password.

The notification does

* Welcome the new member,
* State that the new member will get email,
* Tell the new member how to post,
* Contain a *Reset password* link, and
* Contain an ``Unsubscribe`` ``mailto``.

Resources
=========

- Code repository:
  https://github.com/groupserver/gs.group.member.add.base/
- Translations:
  https://www.transifex.com/projects/p/gs-group-member-add-base/
- Questions and comments to
  http://groupserver.org/groups/development
- Report bugs at https://redmine.iopen.net/projects/groupserver

.. _GroupServer: http://groupserver.org/
.. _GroupServer.org: http://groupserver.org/
.. _OnlineGroups.Net: https://onlinegroups.net
.. _Michael JasonSmith: http://groupserver.org/p/mpj17
.. [#invite] See <https://github.com/groupserver/gs.group.member.invite>
.. [#register] See <https://github.com/groupserver/gs.group.member.signup>
.. [#join] See <https://github.com/groupserver/gs.group.member.join>
.. [#login] In the experience of `OnlineGroups.Net`_ almost all groups are
            secret, with the majority of the remainder being private. The
            public groups are enough of an exception that we do not have to
            worry about them.
.. [#digest] In the corner-case of someone being added to a public group
             and needing *Web only* or *Digest* then the administrator can
             change the email settings of the new member after adding them.


docs/HISTORY.rst
Changelog
=========

3.3.6 (2016-08-09)
------------------

* [DE] Fixing the HTML in the German translation, thanks to
  `Stephan Blendinger`_

.. _Stephan Blendinger:
   https://www.transifex.com/user/profile/stephanblendinger/

3.3.5 (2016-06-20)
------------------

* Fixing the password-link creation in the *Welcome* notification
  — with thanks to Robert_ for spotting `the problem`_
* Dropping some HTML from the plain-text version of the *Welcome*
  notification

.. _Robert: http://groupserver.org/p/1lPEQHbcXYdNuKp41Psm4S
.. _the problem: http://groupserver.org/r/post/1cWESnyfBhWvJHPgEiZl5R

3.3.4 (2015-09-21)
------------------

* Using ``subject`` rather than ``Subject`` in ``mailto:`` URIs
  
3.3.3 (2015-09-15)
------------------

* [DE] Updating the German translation, thanks to Cousin Clara
* [EN] ``s/Dear/Hello/g`` in the *Welcome* notification
* Using the actual support email address, rather than the group
  email address, in the HTML form of the *Welcome* notification,
  thanks to Paul for spotting that

3.3.2 (2015-03-30)
------------------

* Fixing the feedback text on the *Add a member* page

3.3.1 (2015-03-11)
------------------

* [FR] Adding a French translation, thanks to `Razique Mahroua`_

.. _Razique Mahroua: https://www.transifex.com/accounts/profile/Razique/

3.3.0 (2015-03-03)
------------------

* Added internationalisation support
* Adding support for Transifex_
* Following the rename of `gs.group.messages.topic.list`_ from
  ``gs.group.messages.topics``
* Naming the reStructuredText files as such
* Switching to GitHub_ as the primary source-code repository

.. _Transifex:
   https://www.transifex.com/projects/p/gs-group-member-add-base/
.. _gs.group.messages.topic.list:
   https://github.com/groupserver/gs.group.messages.topic.list
.. _GitHub:
   https://github.com/groupserver/gs.group.member.add.base

3.2.0 (2014-06-13)
------------------

* Following the rename `gs.content.form.base` from
  ``gs.content.form``

.. _gs.content.form.base:
   https://github.com/groupserver/gs.content.form.base

3.1.2 (2014-04-25)
------------------

* Metadata update

3.1.1 (2014-02-28)
------------------

* Ensuring the content-type header is reset after sending the
  notifications

3.1.0 (2014-02-20)
------------------

* Switching the base-class of the notifications to those supplied
  by `gs.content.email.base`_

.. _gs.content.email.base:
   https://github.com/groupserver/gs.content.email.base

3.0.2 (2014-02-11)
------------------

* Encoding the text-headers as ASCII strings

3.0.1 (2014-01-30)
------------------

* Verify the address when adding, closing `Bug 4037`_
* Sanitising the email address
* Switching to Unicode literals

.. _Bug 4037: https://redmine.iopen.net/issues/4037

3.0.0 (2014-01-08)
------------------

* Changing the name of the product to `gs.group.member.add.base`_
  from ``gs.group.member.add``

.. _gs.group.member.add.base:
   https://github.com/groupserver/gs.group.member.add.base

2.2.1 (2013-07-17)
------------------

* Adding breadcrumbs
* Cleaning up the JavaScript
* Fixing the ``Content-type`` header

2.2.0 (2013-03-19)
------------------

* Following the changes to the `gs.group.member.base`_ product

.. _gs.group.member.base:
   https://github.com/groupserver/gs.group.member.base


2.1.0 (2013-01-22)
------------------

* Moving the link to the *Add* page to the new *Administer group
  members* viewlet

2.0.1 (2012-11-26)
------------------

* Fixing a Python 2.6 issue with the format strings

2.0.0 (2012-11-13)
------------------

* Updating for the new GroupServer UI
  + Using a full-page layout
  + Creating the *Welcome* notification
  + Creating the *New member* notification
* Changing the base class of the *Add* page to ``GroupForm``
* Splitting the adder code from the *Add* page
* Changing the marker-interface that is used

1.4.0 (2012-06-22)
------------------

* Updating the SQLAlchemy code

1.3.0 (2012-05-15)
------------------

* Following the changes to `gs.group.member.invite.base`_

.. _gs.group.member.invite.base:
   https://github.com/groupserver/gs.group.member.invite.base


1.2.0 (2011-01-26)
------------------

* Adding support for `gs.profile.email.base`_

.. _gs.profile.email.base:
   https://github.com/groupserver/gs.profile.email.base

1.1.0 (2011-01-12)
------------------

* Adding support for `gs.profile.email.verify`_

.. _gs.profile.email.verify:
   https://github.com/groupserver/gs.profile.email.verify

1.0.0 (2010-12-15)
------------------

Initial version. Prior to the creation of this product adding
people was impossible with the standard GroupServer
user-interface.

..  LocalWords:  Changelog reStructuredText GitHub Transifex


version.py
# -*- coding: utf-8 -*-
version = '3.3.6'
release = False

#--------------------------------------------------------------------------#
import sys
if (sys.version_info < (3, )):
    from commands import getstatusoutput
else:
    from subprocess import getstatusoutput  # lint:ok
import datetime
import os
import glob


class CommandError(Exception):
    pass


def execute_command(commandstring):
    status, output = getstatusoutput(commandstring)
    if status != 0:
        m = 'Command "{0}" exited with status {1}'
        msg = m.format(commandstring, status)
        raise CommandError(msg)
    return output


def parse_version_from_package():
    try:
        pkginfo = os.path.join(glob.glob('*.egg-info')[0],
                               'PKG-INFO')
    except:
        pkginfo = ''

    version_string = ''
    if os.path.exists(pkginfo):
        for line in open(pkginfo):
            if line.find('Version: ') == 0:
                version_string = line.strip().split('Version: ')[1].strip()
        if not version_string:
            version_string = '%s-dev' % version
    else:
        version_string = version

    return version_string


def get_version():
    try:
        globalid = execute_command("hg identify -i").strip('+')
        c = "hg log -r %s --template '{date|isodatesec}'" % globalid
        commitdate = execute_command(c)
        # convert date to UTC unix timestamp, using the date command because
        # python date libraries do not stabilise till about 2.6
        dateCommand = 'date -d"%s" --utc +%%s' % commitdate
        timestamp = int(execute_command(dateCommand))

        # finally we have something we can use!
        dt = datetime.datetime.utcfromtimestamp(timestamp)
        datestring = dt.strftime('%Y%m%d%H%M%S')
        if release:
            version_string = version
        else:
            version_string = "%s.dev%s" % (version, datestring)

    except (CommandError, ValueError, TypeError):
        # --=mpj17=-- Usually because we are building out a source-egg,
        # rather than from a Hg source-directory.
        version_string = parse_version_from_package()

    return version_string

if __name__ == '__main__':
    import sys
    sys.stdout.write('{0}\n'.format(get_version()))


gs/group/member/add/base/add.py
# -*- coding: utf-8 -*-
############################################################################
#
# Copyright © 2014, 2015, 2016 OnlineGroups.net and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
############################################################################
from __future__ import absolute_import, unicode_literals
from zope.cachedescriptors.property import Lazy
from zope.formlib import form
from zope.i18n import translate
from Products.Five.browser.pagetemplatefile import ZopeTwoPageTemplateFile
from Products.GSProfile.edit_profile import wym_editor_widget
from gs.core import to_id
from gs.content.form.base import select_widget
from gs.group.base import GroupForm
from gs.group.member.join.notify import (NotifyNewMember as NotifyJoin, NotifyAdmin)
from gs.profile.email.base import EmailUser, sanitise_address
from gs.profile.password.interfaces import IGSPasswordUser
from .addfields import AddFields
from .adder import Adder
from .audit import ADD_NEW_USER, ADD_OLD_USER
from .notifier import Notifier as NotifyAdd
from . import GSMessageFactory as _


class AddEditProfileForm(GroupForm):
    label = _('add-member-title', 'Add a group member')
    pageTemplateFileName = 'browser/templates/edit_profile_add.pt'
    template = ZopeTwoPageTemplateFile(pageTemplateFileName)

    def __init__(self, group, request):
        super(AddEditProfileForm, self).__init__(group, request)

    @Lazy
    def form_fields(self):
        retval = form.Fields(self.addFields.adminInterface,
                             render_context=False)
        retval['tz'].custom_widget = select_widget
        retval['biography'].custom_widget = wym_editor_widget
        return retval

    @Lazy
    def defaultFromEmail(self):
        emailUser = EmailUser(self.context, self.adminInfo)
        retval = emailUser.get_delivery_addresses()[0]
        return retval

    def setUpWidgets(self, ignore_request=False):
        data = {'fromAddr': self.defaultFromEmail}

        siteTz = self.siteInfo.get_property('tz', 'UTC')
        defaultTz = self.groupInfo.get_property('tz', siteTz)
        data['tz'] = defaultTz

        self.widgets = form.setUpWidgets(
            self.form_fields, self.prefix, self.context,
            self.request, form=self, data=data,
            ignore_request=ignore_request)

    @form.action(label=_('add-button', 'Add'), name='add',
                 failure='handle_add_action_failure')
    def handle_add(self, action, data):
        adder = Adder(self.context, self.groupInfo, self.adminInfo)
        toAddr = sanitise_address(data['toAddr'])
        m, userInfo, status = adder.add(toAddr, data)
        msg = translate(m)
        self.status = '<ul>\n{0}\n</ul>'.format(msg)

        # Tell the user
        if status == ADD_NEW_USER:
            notifier = NotifyAdd(self.context, self.request)
            fromAddr = sanitise_address(data['fromAddr'])
            passwd = self.get_password_reset(userInfo, toAddr)
            notifier.notify(self.adminInfo, userInfo, fromAddr, toAddr,
                            passwd)
        elif status == ADD_OLD_USER:
            notifier = NotifyJoin(self.context, self.request)
            notifier.notify(userInfo)

        # Tell the administrator
        if status in (ADD_NEW_USER, ADD_OLD_USER):
            adminNotifier = NotifyAdmin(self.context, self.request)
            for adminInfo in self.groupInfo.group_admins:
                adminNotifier.notify(adminInfo, userInfo)
        assert self.status

    def handle_add_action_failure(self, action, data, errors):
        if len(errors) == 1:
            s = _('single-error', 'There is an error:')
        else:
            s = _('multiple-errors', 'There are errors:')
        self.status = '<p>{0}</p>'.format(translate(s))

    @Lazy
    def addFields(self):
        retval = AddFields(self.context)
        return retval

    @Lazy
    def adminInfo(self):
        retval = self.loggedInUser
        return retval

    @Lazy
    def adminWidgets(self):
        return self.addFields.get_admin_widgets(self.widgets)

    @Lazy
    def profileWidgets(self):
        return self.addFields.standard_widgets(self.widgets)

    @Lazy
    def requiredProfileWidgets(self):
        widgets = self.addFields.get_standard_widgets(self.widgets)
        widgets = [widget for widget in widgets if widget.required]
        return widgets

    @Lazy
    def optionalProfileWidgets(self):
        widgets = self.addFields.get_standard_widgets(self.widgets)
        widgets = [widget for widget in widgets if not(widget.required)]
        return widgets

    def get_password_reset(self, userInfo, email):
        resetId = to_id(email + userInfo.name + self.siteInfo.name)

        passwordUser = IGSPasswordUser(userInfo)
        passwordUser.add_password_reset(resetId)

        u = '{siteUrl}/r/password/{resetId}'
        retval = u.format(siteUrl=self.siteInfo.url, resetId=resetId)
        assert retval
        return retval



gs/group/member/add/base/audit.py
# -*- coding: utf-8 -*-
############################################################################
#
# Copyright © 2014 OnlineGroups.net and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
############################################################################
from __future__ import unicode_literals
from pytz import UTC
from datetime import datetime
from zope.component.interfaces import IFactory
from zope.interface import implements, implementedBy
from Products.CustomUserFolder.userinfo import userInfo_to_anchor
from Products.GSGroup.groupInfo import groupInfo_to_anchor
from Products.GSAuditTrail import (IAuditEvent, BasicAuditEvent,
                                   AuditQuery, event_id_from_data)

SUBSYSTEM = 'gs.group.member.add'
import logging
log = logging.getLogger(SUBSYSTEM)

UNKNOWN = '0'
ADD_NEW_USER = '1'
ADD_OLD_USER = '2'
ADD_EXISTING_MEMBER = '3'


class AuditEventFactory(object):
    implements(IFactory)

    title = 'User Profile Add Audit-Event Factory'
    description = 'Creates a GroupServer audit event for user add'

    def __call__(self, context, event_id, code, date, userInfo,
                 instanceUserInfo, siteInfo, groupInfo, instanceDatum='',
                 supplementaryDatum='', subsystem=''):

        if (code == ADD_NEW_USER):
            event = AddNewUserEvent(context, event_id, date, userInfo,
                                    instanceUserInfo, siteInfo, groupInfo,
                                    instanceDatum, supplementaryDatum)
        elif (code == ADD_OLD_USER):
            event = AddOldUserEvent(context, event_id, date, userInfo,
                                    instanceUserInfo, siteInfo, groupInfo,
                                    instanceDatum, supplementaryDatum)
        elif (code == ADD_EXISTING_MEMBER):
            event = AddExistingMemberEvent(context, event_id, date,
                                           userInfo, instanceUserInfo,
                                           siteInfo, groupInfo,
                                           instanceDatum,
                                           supplementaryDatum)
        else:
            event = BasicAuditEvent(context, event_id, UNKNOWN, date,
                                    userInfo, instanceUserInfo, siteInfo,
                                    groupInfo, instanceDatum,
                                    supplementaryDatum, SUBSYSTEM)
        assert event
        return event

    def getInterfaces(self):
        return implementedBy(BasicAuditEvent)


class AddNewUserEvent(BasicAuditEvent):
    """Administrator adding a New User Event.

    The "instanceDatum" is the address used to create the new user.
    """
    implements(IAuditEvent)

    def __init__(self, context, id, d, userInfo, instanceUserInfo, siteInfo,
                 groupInfo, instanceDatum, supplementaryDatum):

        BasicAuditEvent.__init__(self, context, id, ADD_NEW_USER, d,
                                 userInfo, instanceUserInfo, siteInfo,
                                 groupInfo, instanceDatum,
                                 supplementaryDatum, SUBSYSTEM)

    def __unicode__(self):
        r = 'Administrator {0} ({1}) adding a new user {2} ({3}) '\
            'with address <{4}> to join {5} ({6}) on {7} ({8})'
        retval = r.format(self.userInfo.name, self.userInfo.id,
                          self.instanceUserInfo.name,
                          self.instanceUserInfo.id, self.instanceDatum,
                          self.groupInfo.name, self.groupInfo.id,
                          self.siteInfo.name, self.siteInfo.id)
        return retval

    @property
    def xhtml(self):
        cssClass = 'audit-event profile-add-event-{0}'.format(self.code)
        email = '<code class="email">{}</code>'.format(self.instanceDatum)
        m = '<span class="{cssClass}">Adding the new user {user} (with '\
            'the email address {email}) to join {group}.</span>'
        retval = m.format(cssClass=cssClass, email=email,
                          user=userInfo_to_anchor(self.instanceUserInfo),
                          group=groupInfo_to_anchor(self.groupInfo))
        if ((self.instanceUserInfo.id != self.userInfo.id)
                and not(self.userInfo.anonymous)):
            u = userInfo_to_anchor(self.userInfo)
            retval = '{0} &#8212; {1}'.format(retval, u)
        return retval


class AddOldUserEvent(BasicAuditEvent):
    """Administrator adding an old User Event.

    The "instanceDatum" is the address used to match the old user.
    """
    implements(IAuditEvent)

    def __init__(self, context, id, d, userInfo, instanceUserInfo,
                 siteInfo, groupInfo, instanceDatum, supplementaryDatum):

        BasicAuditEvent.__init__(self, context, id,
          ADD_OLD_USER, d, userInfo, instanceUserInfo,
          siteInfo, groupInfo, instanceDatum, supplementaryDatum,
          SUBSYSTEM)

    def __unicode__(self):
        retval = 'Administrator %s (%s) adding an existing user '\
                  '%s (%s) with address <%s> to join %s (%s) on %s (%s)' %\
          (self.userInfo.name, self.userInfo.id,
          self.instanceUserInfo.name, self.instanceUserInfo.id,
          self.instanceDatum,
          self.groupInfo.name, self.groupInfo.id,
          self.siteInfo.name, self.siteInfo.id)
        return retval

    @property
    def xhtml(self):
        cssClass = 'audit-event profile-add-event-{}'.format(self.code)
        m = '<span class="{cssClass}">Adding the existing user {user} '\
            'to {group}</span>'
        retval = m.format(cssClass=cssClass,
                            user=userInfo_to_anchor(self.instanceUserInfo),
                            group=groupInfo_to_anchor(self.groupInfo))
        if ((self.instanceUserInfo.id != self.userInfo.id)
            and not(self.userInfo.anonymous)):
            retval = '{} &#8212; {}'.format(retval,
                                            userInfo_to_anchor(self.userInfo))
        return retval


class AddExistingMemberEvent(BasicAuditEvent):
    """Administrator Adding an Existing Group Member.

    The "instanceDatum" is the address used to match the existing group
    member.
    """
    implements(IAuditEvent)

    def __init__(self, context, id, d, userInfo, instanceUserInfo,
        siteInfo, groupInfo, instanceDatum, supplementaryDatum):

        BasicAuditEvent.__init__(self, context, id,
          ADD_OLD_USER, d, userInfo, instanceUserInfo,
          siteInfo, groupInfo, instanceDatum, supplementaryDatum,
          SUBSYSTEM)

    def __unicode__(self):
        retval = 'Administrator %s (%s) tried to add an existing '\
                  'group member %s (%s) with address <%s> to join %s (%s) '\
                  'on %s (%s)' %\
          (self.userInfo.name, self.userInfo.id,
          self.instanceUserInfo.name, self.instanceUserInfo.id,
          self.instanceDatum,
          self.groupInfo.name, self.groupInfo.id,
          self.siteInfo.name, self.siteInfo.id)
        return retval

    @property
    def xhtml(self):
        cssClass = 'audit-event profile-add-event-%s' % self.code
        retval = '<span class="%s">Tried to add the existing member '\
            ' %s to join %s.</span>' %\
            (cssClass, userInfo_to_anchor(self.instanceUserInfo),
            groupInfo_to_anchor(self.groupInfo))
        if ((self.instanceUserInfo.id != self.userInfo.id)
            and not(self.userInfo.anonymous)):
            retval = '%s &#8212; %s' %\
              (retval, userInfo_to_anchor(self.userInfo))
        return retval


class Auditor(object):
    def __init__(self, siteInfo, groupInfo, adminInfo, userInfo):
        self.siteInfo = siteInfo
        self.groupInfo = groupInfo
        self.adminInfo = adminInfo
        self.userInfo = userInfo

        self.queries = AuditQuery()

        self.factory = AuditEventFactory()

    def info(self, code, instanceDatum='', supplementaryDatum=''):
        d = datetime.now(UTC)
        eventId = event_id_from_data(self.adminInfo, self.userInfo,
            self.siteInfo, code, instanceDatum, supplementaryDatum)

        e = self.factory(self.userInfo.user, eventId, code, d,
                self.adminInfo, self.userInfo, self.siteInfo,
                self.groupInfo, instanceDatum, supplementaryDatum,
                SUBSYSTEM)

        self.queries.store(e)
        log.info(e)


gs/group/member/add/base/notifier.py
# -*- coding: utf-8 -*-
############################################################################
#
# Copyright © 2014 OnlineGroups.net and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
############################################################################
from __future__ import unicode_literals
from zope.i18n import translate
from gs.content.email.base import GroupNotifierABC
from gs.profile.notify import MessageSender
from . import GSMessageFactory as _


class Notifier(GroupNotifierABC):
    htmlTemplateName = 'gs-group-member-add-welcome.html'
    textTemplateName = 'gs-group-member-add-welcome.txt'

    def notify(self, adminInfo, userInfo, fromAddr, toAddr, passwordLink):
        subject = _('welcome-subject', 'Welcome to ${groupName}',
                    mapping={'groupName': self.groupInfo.name})
        translatedSubject = translate(subject)
        text = self.textTemplate(adminInfo=adminInfo, userInfo=userInfo)
        html = self.htmlTemplate(adminInfo=adminInfo, userInfo=userInfo,
                                 passwordLink=passwordLink)

        sender = MessageSender(self.context, userInfo)
        sender.send_message(translatedSubject, text, html, fromAddr,
                            [toAddr])
        self.reset_content_type()


gs/group/member/add/base/adder.py
# -*- coding: utf-8 -*-
############################################################################
#
# Copyright © 2014 OnlineGroups.net and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
############################################################################
from __future__ import absolute_import, unicode_literals
from zope.component import createObject
from zope.formlib import form
from zope.i18n import translate
from gs.group.member.base.utils import user_member_of_group
from gs.group.member.join.interfaces import IGSJoiningUser
from gs.profile.email.base import NewEmailAddress, EmailAddressExists, \
    EmailUser
from gs.profile.email.verify import EmailVerificationUser
from Products.CustomUserFolder.userinfo import userInfo_to_anchor
from Products.GSGroup.groupInfo import groupInfo_to_anchor
from Products.GSProfile.utils import create_user_from_email, \
    enforce_schema
from .addfields import AddFields
from .audit import Auditor, ADD_NEW_USER, ADD_OLD_USER, ADD_EXISTING_MEMBER
from . import GSMessageFactory as _


class Adder(object):
    def __init__(self, context, groupInfo, adminInfo):
        self.context = context
        self.groupInfo = groupInfo
        self.adminInfo = adminInfo

    def add(self, toAddr, profileDict):
        emailChecker = NewEmailAddress(title='Email')
        emailChecker.context = self.context
        try:
            emailChecker.validate(toAddr)
        except EmailAddressExists:
            msg, userInfo, status = self.add_existing_user(toAddr)
        else:
            msg, userInfo, status = self.add_new_user(toAddr, profileDict)
        return (msg, userInfo, status)

    def add_existing_user(self, toAddr):
        acl_users = self.context.acl_users
        user = acl_users.get_userByEmail(toAddr)
        if not user:
            m = 'User for address <{0}> not found'.format(toAddr)
            raise LookupError(m)
        # get the user object in the context of the group and site
        userInfo = createObject('groupserver.UserFromId', self.context,
                                user.getId())
        auditor = self.get_auditor(userInfo)
        e = '<code class="email">{0}</code>'.format(toAddr)
        if user_member_of_group(user, self.groupInfo):
            status = ADD_EXISTING_MEMBER
            auditor.info(status, toAddr)
            m = _('existing-member-msg',
                  '<li>The person with the email address ${email} '
                  '&#8213; ${user} &#8213; is already a member of '
                  '${group}. No changes to the profile of ${user} have '
                  'been made.</li>',
                  mapping={'email': e,
                           'user': userInfo_to_anchor(userInfo),
                           'group': groupInfo_to_anchor(self.groupInfo)})
        else:
            status = ADD_OLD_USER
            auditor.info(status, toAddr)
            joininguser = IGSJoiningUser(userInfo)
            joininguser.silent_join(self.groupInfo)

            # Verify the address <https://redmine.iopen.net/issues/4037>
            eu = EmailUser(self.context, userInfo)
            if not eu.is_address_verified(toAddr):
                evu = EmailVerificationUser(self.context, userInfo, toAddr)
                verificationId = evu.create_verification_id()
                evu.add_verification_id(verificationId)
                evu.verify_email(verificationId)

            m = _('existing-profile-msg',
                  '<li>Adding the existing participant with  the email '
                  'address ${email} &#8213; ${user} &#8213; to '
                  '${group}.</li>',
                  mapping={'email': e,
                           'user': userInfo_to_anchor(userInfo),
                           'group': groupInfo_to_anchor(self.groupInfo)})
        msg = translate(m)
        retval = (msg, userInfo, status)
        return retval

    def add_new_user(self, toAddr, profileDict):
        # Email address does not exist, but it is a legitimate address
        user = create_user_from_email(self.context, toAddr)
        # force verify
        vid = '%s-%s-verified' % (toAddr, self.adminInfo.id)
        evu = createObject('groupserver.EmailVerificationUserFromEmail',
                           self.context, toAddr)
        evu.add_verification_id(vid)
        evu.verify_email(vid)

        # get the user object in the context of the group and site
        userInfo = createObject('groupserver.UserFromId', self.context,
                                user.id)
        self.add_profile_attributes(userInfo, profileDict)
        auditor = self.get_auditor(userInfo)
        status = ADD_NEW_USER
        auditor.info(status, toAddr)
        joininguser = IGSJoiningUser(userInfo)
        joininguser.silent_join(self.groupInfo)
        e = '<code class="email">{0}</code>'.format(toAddr)
        msg = _('new-profile-member',
                '<li>A profile for ${user} has been created, and given the '
                'email address ${email}.</li>\n'
                '<li>${user} has been joined to ${group}.</li>\n',
                mapping={'email': e,
                         'user': userInfo_to_anchor(userInfo),
                         'group': groupInfo_to_anchor(self.groupInfo)})
        retval = (msg, userInfo, status)
        return retval

    def add_profile_attributes(self, userInfo, data):
        addFields = AddFields(self.context)
        enforce_schema(userInfo.user, addFields.profileInterface)
        f = form.Fields(addFields.adminInterface, render_context=False)
        fields = f.select(*addFields.profileFieldIds)
        for field in fields:
            field.interface = addFields.profileInterface
        form.applyChanges(userInfo.user, fields, data)

    def get_auditor(self, userInfo):
        auditor = Auditor(self.groupInfo.siteInfo, self.groupInfo,
                          self.adminInfo, userInfo)
        return auditor
